File: apps/mobile/shared/Sources/FlagshipCore/ServerTransferOrders.swift
import Foundation
import CryptoKit

/// Transfer-a-box orders — the Swift mirror of the `server-transfer-offer` +
/// `server-transfer-claim` envelopes in `packages/protocol/src/legacyEnvelopes.ts`
/// (docs/account-deletion-and-name-reclaim.md §4).
///
/// Cross-account ownership handoff, two parties / two envelopes:
///  - `server-transfer-offer` is minted by the CURRENT owner's phone (giver IRK)
///    and encoded into the QR shown on the box's detail page. It does NOT name
///    the acquirer (unknown until they scan); it commits only to the box + a
///    one-time short-TTL nonce.
///  - `server-transfer-claim` is minted by the ACQUIRER's phone (acquirer IRK)
///    after scanning, binding the acquirer's identity (username + IRK pub) to
///    the offer's nonce so `.com` moves ownership to a SPECIFIC new account.
///
/// Canonical bytes (byte-identical to TS + the Kotlin mirror):
///
///   flagship/server-transfer-offer/v1|<serverDomain>|<transferNonce>|<issuedAt>|<expiresAt>
///   flagship/server-transfer-claim/v2|<serverDomain>|<transferNonce>|<acquirerUsername>|<acquirerIrkPubHex>|<acquirerAdminRootPubHex or "">|<issuedAt>
///
/// `serverDomain`, `transferNonce`, and `acquirerUsername` are lowercased into
/// the canonical bytes (matching the TS `.toLowerCase()`).
public struct ServerTransferOfferOrder: Equatable, Sendable {
    public static let canonicalTag = "flagship/server-transfer-offer/v1"

    public let serverDomain: String
    public let transferNonce: String
    public let issuedAt: Int64
    public let expiresAt: Int64

    public init(serverDomain: String, transferNonce: String, issuedAt: Int64, expiresAt: Int64) {
        self.serverDomain = serverDomain
        self.transferNonce = transferNonce
        self.issuedAt = issuedAt
        self.expiresAt = expiresAt
    }

    public func canonicalBytes() -> Data {
        Data(
            [
                Self.canonicalTag,
                serverDomain.lowercased(),
                transferNonce.lowercased(),
                String(issuedAt),
                String(expiresAt),
            ].joined(separator: "|").utf8
        )
    }

    public func sign(with irk: Curve25519.Signing.PrivateKey) throws -> Data {
        try irk.signature(for: canonicalBytes())
    }
}

public struct ServerTransferClaimOrder: Equatable, Sendable {
    /// v2 (spec §9.8): the claim canonical now COMMITS to the acquirer's admin
    /// master root pub, so `.com` can't substitute a different anchor for the
    /// box to re-pin — the acquirer's own signature covers it. Empty string ⇒
    /// the acquirer account has no admin root (legacy).
    public static let canonicalTag = "flagship/server-transfer-claim/v2"

    public let serverDomain: String
    public let transferNonce: String
    public let acquirerUsername: String
    /// The acquirer's owner IRK pubkey, hex — ownership re-binds to this.
    public let acquirerIrkPubHex: String
    /// The acquirer's admin master root pubkey, hex — the anchor the box
    /// re-pins on re-home. "" when the acquirer account has no admin root.
    public let acquirerAdminRootPubHex: String
    public let issuedAt: Int64

    public init(serverDomain: String, transferNonce: String, acquirerUsername: String, acquirerIrkPubHex: String, acquirerAdminRootPubHex: String = "", issuedAt: Int64) {
        self.serverDomain = serverDomain
        self.transferNonce = transferNonce
        self.acquirerUsername = acquirerUsername
        self.acquirerIrkPubHex = acquirerIrkPubHex
        self.acquirerAdminRootPubHex = acquirerAdminRootPubHex
        self.issuedAt = issuedAt
    }

    public func canonicalBytes() -> Data {
        Data(
            [
                Self.canonicalTag,
                serverDomain.lowercased(),
                transferNonce.lowercased(),
                acquirerUsername.lowercased(),
                acquirerIrkPubHex.lowercased(),
                acquirerAdminRootPubHex.lowercased(),
                String(issuedAt),
            ].joined(separator: "|").utf8
        )
    }

    public func sign(with irk: Curve25519.Signing.PrivateKey) throws -> Data {
        try irk.signature(for: canonicalBytes())
    }
}

/// v1-sec GAP 3 — the LEGACY (no-admin-root) re-home authorization. Swift mirror
/// of the spine's `flagship/server-rehome-auth/v1` envelope; the canonical bytes
/// MUST match the TS `signRehomeAuthorization` byte-for-byte or the box-side
/// verify against its pinned owner IRK fails.
///
/// A box with NO pinned admin master root re-homes ONLY on this proof, signed by
/// the GIVER's owner IRK (the box's config-pinned owner IRK until it re-homes),
/// naming the acquirer explicitly. The box verifies it against its pin before
/// writing the re-home marker; `.com` relays it but cannot forge it. The
/// admin-tier path keeps its stronger `AdminRootTransfer` proof instead.
///
/// Canonical bytes (`|`-separated), matching TS `hex()` (lowercase) exactly:
///   flagship/server-rehome-auth/v1 | oldServerDomain | newServerDomain
///     | acquirerIrkPubHex | issuedAt
public struct RehomeAuthorizationOrder: Equatable, Sendable {
    public static let canonicalTag = "flagship/server-rehome-auth/v1"

    /// The box's OLD canonical FQDN (`<server>.<giver>.<apex>`).
    public let oldServerDomain: String
    /// The NEW canonical FQDN to re-home to (`<server>.<acquirer>.<apex>`).
    public let newServerDomain: String
    /// The acquirer's owner-IRK pubkey, lowercase hex — ownership re-binds to this.
    public let acquirerIrkPubHex: String
    public let issuedAt: Int64

    public init(oldServerDomain: String, newServerDomain: String, acquirerIrkPubHex: String, issuedAt: Int64) {
        self.oldServerDomain = oldServerDomain
        self.newServerDomain = newServerDomain
        self.acquirerIrkPubHex = acquirerIrkPubHex
        self.issuedAt = issuedAt
    }

    public func canonicalBytes() -> Data {
        Data(
            [
                Self.canonicalTag,
                oldServerDomain.lowercased(),
                newServerDomain.lowercased(),
                acquirerIrkPubHex.lowercased(),
                String(issuedAt),
            ].joined(separator: "|").utf8
        )
    }

    /// Sign with the GIVER's owner IRK (the anchor the box pins today).
    public func sign(withGiverIrk key: Curve25519.Signing.PrivateKey) throws -> Data {
        try key.signature(for: canonicalBytes())
    }

    /// Verify against the giver's owner-IRK pubkey (the box's pinned anchor).
    /// Returns false (never throws) on a malformed key/signature.
    public func verify(signature: Data, giverIrkPub: Data) -> Bool {
        guard let pub = try? Curve25519.Signing.PublicKey(rawRepresentation: giverIrkPub) else {
            return false
        }
        return pub.isValidSignature(signature, for: canonicalBytes())
    }
}

File: apps/mobile/android/app/src/main/java/com/flagshipserver/app/core/ServerTransferOrders.kt
// Transfer-a-box orders — the Kotlin mirror of iOS
// `FlagshipCore/ServerTransferOrders.swift` and the `server-transfer-offer` +
// `server-transfer-claim` envelopes in `packages/protocol/src/legacyEnvelopes.ts`
// (docs/account-deletion-and-name-reclaim.md §4).
//
// Cross-account ownership handoff, two parties / two envelopes:
//  - server-transfer-offer: minted by the CURRENT owner's phone (giver IRK),
//    encoded into the QR on the box detail page. Does NOT name the acquirer.
//  - server-transfer-claim: minted by the ACQUIRER's phone (acquirer IRK) after
//    scanning, binding the acquirer's username + IRK pub to the offer's nonce.
//
// Canonical bytes (byte-identical to TS + the Swift mirror):
//
//   flagship/server-transfer-offer/v1|<serverDomain>|<transferNonce>|<issuedAt>|<expiresAt>
//   flagship/server-transfer-claim/v2|<serverDomain>|<transferNonce>|<acquirerUsername>|<acquirerIrkPubHex>|<acquirerAdminRootPubHex>|<issuedAt>
//
// serverDomain, transferNonce, acquirerUsername, and the pub hexes are
// lowercased into the canonical bytes (matching the TS `.toLowerCase()`).
//
// Claim v2 (device-admin-tier spec §9.8): the acquirer's ADMIN MASTER ROOT pub
// rides INSIDE the signed canonical (empty string when the acquirer account has
// none) so the box can re-pin the acquirer's authority anchor at re-home — a
// value `.com` merely relayed outside the signature could be swapped.

package com.flagshipserver.app.core

object ServerTransferOfferOrder {
    const val CANONICAL_TAG = "flagship/server-transfer-offer/v1"

    fun canonicalBytes(serverDomain: String, transferNonce: String, issuedAt: Long, expiresAt: Long): ByteArray =
        listOf(
            CANONICAL_TAG,
            serverDomain.lowercase(),
            transferNonce.lowercase(),
            issuedAt.toString(),
            expiresAt.toString(),
        ).joinToString("|").toByteArray(Charsets.UTF_8)
}

object ServerTransferClaimOrder {
    const val CANONICAL_TAG = "flagship/server-transfer-claim/v2"

    fun canonicalBytes(
        serverDomain: String,
        transferNonce: String,
        acquirerUsername: String,
        acquirerIrkPubHex: String,
        acquirerAdminRootPubHex: String,
        issuedAt: Long,
    ): ByteArray =
        listOf(
            CANONICAL_TAG,
            serverDomain.lowercase(),
            transferNonce.lowercase(),
            acquirerUsername.lowercase(),
            acquirerIrkPubHex.lowercase(),
            acquirerAdminRootPubHex.lowercase(),
            issuedAt.toString(),
        ).joinToString("|").toByteArray(Charsets.UTF_8)
}

// v1-sec GAP 3 — the LEGACY (no-admin-root) re-home authorization. Kotlin mirror
// of the spine's `flagship/server-rehome-auth/v1` envelope + the Swift
// `RehomeAuthorizationOrder`. A box with NO pinned admin master root re-homes
// ONLY on this proof, signed by the GIVER's owner IRK (the box's pinned owner
// IRK until it re-homes), naming the acquirer explicitly. The box verifies it
// against its pin before writing the re-home marker; `.com` relays but cannot
// forge it.
//
// Canonical bytes (byte-identical to TS + the Swift mirror):
//
//   flagship/server-rehome-auth/v1|<oldServerDomain>|<newServerDomain>|<acquirerIrkPubHex>|<issuedAt>
//
// The domains + pub hex are lowercased into the canonical bytes.
object RehomeAuthorizationOrder {
    const val CANONICAL_TAG = "flagship/server-rehome-auth/v1"

    fun canonicalBytes(
        oldServerDomain: String,
        newServerDomain: String,
        acquirerIrkPubHex: String,
        issuedAt: Long,
    ): ByteArray =
        listOf(
            CANONICAL_TAG,
            oldServerDomain.lowercase(),
            newServerDomain.lowercase(),
            acquirerIrkPubHex.lowercase(),
            issuedAt.toString(),
        ).joinToString("|").toByteArray(Charsets.UTF_8)
}
